File: week-03/day-3/src/pirates/Ship.java
package pirates;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Ship {
    String name;
    Pirate captain;
    List<Pirate> crew = new ArrayList<>();

    public Ship(String name) {
        this.name = name;
    }

    public void fillShip(String nameOfCaptain) {
        captain = new Pirate(nameOfCaptain);
        int randomNumber = new Random().nextInt(50) + 1;
        for (int i = 0; i < randomNumber; i++) {
            crew.add(new Pirate());
        }
    }

    public void introduceYourCaptainAndCrew() {
        System.out.println("The captain of the " + name + " is currently " + captain.name + ".");
        if (captain.isAlive == false) {
            System.out.println("The captain is dead.");
        } else {
            System.out.println("Our captain " + captain.name + " has drunk " + captain.levelOfIntoxication + " rounds of rum today.");
            if (captain.hasAParrot == true) {
                System.out.println("Captain has a parrot.");
            }
            if (captain.isPassedOut == true) {
                System.out.println("Captain " + captain.name + " has passed out in a brawl / from heavy drinking.");
            } else {
                System.out.println("Captain " + captain.name + " is vital and unharmed.");
            }

        }

        int aliveMembersOfTheCrew = crew.size();
        for (Pirate memberOfTheCrew : crew) {
            if (memberOfTheCrew.isAlive) {
                aliveMembersOfTheCrew++;
            }

        }

        System.out.println("The " + name + " has a fine crew of " + aliveMembersOfTheCrew + " alive pirates.");

    }

    public boolean battle (Ship anotherShip) {
        if (calculateBattleScore(anotherShip)) {
            captain.drinkSomeRum();
            int numberOfRumsForCaptain = new Random().nextInt(10);

            for (int i = 0; i < numberOfRumsForCaptain; i++) {
                captain.drinkSomeRum();
            }

            for (Pirate memberOfTheCrew : crew) {
                int numberOfRumsForCrew = new Random().nextInt(10);
                for (int i = 0; i < numberOfRumsForCrew; i++) {
                    if (memberOfTheCrew.isAlive) {
                        memberOfTheCrew.drinkSomeRum();
                    }
                }
            }

            int numberOfDeaths = new Random().nextInt(anotherShip.crew.size() + 1);
            int indexOfPirate = 0;
            while (numberOfDeaths > 0 && indexOfPirate < anotherShip.crew.size()) {
                if (anotherShip.crew.get(indexOfPirate).isAlive) {
                    anotherShip.crew.get(indexOfPirate).die();
                }
                indexOfPirate++;
                numberOfDeaths--;
            }

            return true;

        } else {
            anotherShip.captain.drinkSomeRum();
            int numberOfRumsForCaptain = new Random().nextInt(10);

            for (int i = 0; i < numberOfRumsForCaptain; i++) {
                anotherShip.captain.drinkSomeRum();
            }

            for (Pirate memberOfTheCrew : anotherShip.crew) {
                int numberOfRumsForCrew = new Random().nextInt(10);
                for (int i = 0; i < numberOfRumsForCrew; i++) {
                    if (memberOfTheCrew.isAlive) {
                        memberOfTheCrew.drinkSomeRum();
                    }
                }
            }

            int numberOfDeaths = new Random().nextInt(crew.size() + 1);
            int indexOfPirate = 0;
            while (numberOfDeaths > 0 && indexOfPirate < crew.size()) {
                if (crew.get(indexOfPirate).isAlive) {
                    crew.get(indexOfPirate).die();
                }
                indexOfPirate++;
                numberOfDeaths--;
            }

            return false;
        }
    }

    public boolean calculateBattleScore(Ship anotherShip) {
        int scoreOfThisShip = 0;
        int scoreOfAnotherShip = 0;

        for (Pirate memberOfTheCrew : crew) {
            if (memberOfTheCrew.isAlive) {
                scoreOfThisShip++;
            }
        }

        scoreOfThisShip -= captain.levelOfIntoxication;
        if (captain.isPassedOut) {
            scoreOfThisShip -= 15;
        }
        if (!captain.isAlive) {
            scoreOfThisShip = 0;
        }


        for (Pirate memberOfTheCrew : anotherShip.crew) {
            if (memberOfTheCrew.isAlive) {
                scoreOfAnotherShip++;
            }
        }

        scoreOfAnotherShip -= anotherShip.captain.levelOfIntoxication;
        if (anotherShip.captain.isPassedOut) {
            scoreOfThisShip -= 15;
        }
        if (!anotherShip.captain.isAlive) {
            scoreOfThisShip = 0;
        }


        if (scoreOfThisShip >= scoreOfAnotherShip) {
            return true;
        } else {
            return false;
        }
    }


}

File: week-10-project-phase/day-4/Pirates/Ship.cs
using System;
using System.Collections.Generic;

namespace Pirates
{
    class Ship
    {
        private List<Pirate> crew;
        internal List<Pirate> Crew { get => crew; set => crew = value; }

        private Pirate captain;
        internal Pirate Captain { get => captain; set => captain = value; }

        public Ship()
        {
            this.Crew = new List<Pirate>();
            this.FillShip();
        }

        public void FillShip()
        {
            this.Captain = new Pirate();
            for (int i = 0; i < new Random().Next(50) + 1; i++)
            {
                Crew.Add(new Pirate());
            }
        }

        public void IntroduceYourCaptainAndCrew()
        {
            Console.WriteLine($"The captain has drunk {this.Captain.LevelOfIntoxication} rounds of rum today.");
            Console.WriteLine(this.Captain.IsDead ? "The captain is dead." : this.Captain.IsPassedOut ? "The captain has passed out." : "The captain is in really good shape.");
            Console.WriteLine($"The ship has {GetAlivePirates().Count} alive members in the crew.");
        }

        public List<Pirate> GetAlivePirates()
        {
            List<Pirate> alivePirates = new List<Pirate>();
            foreach (Pirate pirate in Crew)
            {
                if(!pirate.IsDead)
                {
                    alivePirates.Add(pirate);
                }
            }
            return alivePirates;
        }

        public bool Battle(Ship anotherShip)
        {
            if(this.CalculateScore() >= anotherShip.CalculateScore())
            {
                anotherShip.HaveAParty();
                this.SufferLosses(new Random().Next(this.Crew.Count));
                return false;
            } else
            {
                this.HaveAParty();
                anotherShip.SufferLosses(new Random().Next(this.Crew.Count));
                return true;
            }

        }

        public int CalculateScore()
        {
            return this.GetAlivePirates().Count - this.Captain.LevelOfIntoxication;
        }

        public void SufferLosses(int numberOfDeaths)
        {
            for (int i = 0; i < numberOfDeaths; i++)
            {
                if (!Crew[i].IsDead)
                {
                    Crew[i].Die();
                }
                else
                {
                    numberOfDeaths++;
                }
            }
        }

        public void HaveAParty()
        {
            for (int i = 0; i < new Random().Next(5); i++)
            {
                this.Captain.DrinkSomeRum();
            }

            foreach (Pirate pirate in Crew)
            {
                for (int i = 0; i < new Random().Next(5); i++)
                {
                    pirate.DrinkSomeRum();
                }
            }
        }
    }
}
